File: chap5/GaussianIntegral.f90
!// 注: 参考来源http://fcode.cn/algorithm-73-1.html  
Module Gauss_Legendre !//高斯—勒让德积分高斯点及权重的求解模块
  Implicit none
  Integer, parameter :: n  = 11                          !// 设置求解高斯点的个数
  Integer, parameter :: DP = selected_real_kind( p=13 )  !// 设置kind的数值
  Real(kind=DP), parameter :: eps = 1.0e-15_DP           !// 精度设置
Contains
  Real(Kind=DP) function N_Legendre(x) !// 生成n阶勒让德多项式
    Implicit none
    Integer :: i
    Real(Kind=DP) :: a(n), x
    a(1) = x !// 1阶勒让德多项式
    a(2) = 1.5_DP*x*x - 0.5_DP !// 2阶勒让德多项式
    Do i = 3, n
      a(i) = ( dble(i+i-1)*x*a(i-1) - dble(i-1)*a(i-2) ) / dble(i) !// 利用递推关系产生n阶勒让德多项式
    End do
    N_Legendre=a(n) !//生成的n阶勒让德多项式
  End function N_Legendre

  Real(Kind=DP) Function N1_Legendre(x)  !// 生成n-1阶勒让德多项式 
    Implicit none
    Integer :: i
    Real (Kind=DP) :: a(n), x
    a(1) = x
    a(2) = 1.5_DP*x**2 - 0.5_DP
    Do i = 3, n - 1
      a(i) = (2*i-1)*x*a(i-1)/i - (i-1)*a(i-2)/i
    End Do
    N1_Legendre = a(n-1)     
  End function N1_Legendre
  
  Real(Kind=DP) function DN_Legendre(x)  !// 生成n阶勒让德多项式的导数表达式
    Implicit none
    Integer :: i
    Real(Kind=DP) :: a(n), x
    a(1) = x  !// 1阶勒让德多项式
    a(2) = 1.5_DP*x*x - 0.5_DP !// 2阶勒让德多项式
    Do i = 3, n
      a(i) = ( dble(i+i-1)*x*a(i-1) - dble(i-1)*a(i-2) ) / dble(i) !// 利用递推关系产生n阶勒让德多项式
    End Do
    DN_Legendre = ( a(n-1) - x*a(n) )*dble(n) / (1.0_DP - x*x ) 
  End function DN_Legendre

  Real(Kind=DP) function NewtonIteration(a, b) !// 牛顿法求解函数的解
    Implicit none
    Integer :: i
    Real(Kind=DP) :: a, b, x, xtmp
    Integer, parameter :: nloop = 2000
    !// a,b是传递进来的划分好的有一个解存在的区间
    x = ( a + b ) / 2.d0  !// 初始估计值
    i = 0
    Do 
      xtmp = x - N_Legendre(x) / DN_Legendre(x)   !// X(i+1) = Xi - f(Xi) / f'(Xi)  i = 1,2,...N
      i = i + 1
      If ( abs( xtmp-x ) < eps .and. i > nloop ) exit
      x = xtmp
    End do 
    NewtonIteration = x
  End function NewtonIteration

  Subroutine root_coeff ( f_root, f_coeff )  !// 计算N阶勒让德多项式的根与去做权重系数
    Implicit none
    Real(Kind=DP) :: m, nstep, f_root(n), f_coeff(n) !// 定义数组,大小n由module开始声明。
    Integer :: i, j
    Real(kind=DP), parameter :: h = 1.d-6
    j = 0   !// 赋值控制循环变量的初值           
    m = -1.d0 + h   !// 设置计算域[-1，1] 的下限，即代替-1 
    nstep = nint(2.d0/h)
    Do i = 1, nstep   !// 这个循环次数应该是由步长0.000001决 定,计算方法：2000000=2/0.000001     
      If ( N_Legendre(m)*N_Legendre(m+h) < 0 ) then   !// 从下限处开始往上逐步累加
        j = j + 1    !// 记录这是第几个解
        f_root(j) = NewtonIteration( m, m+h )!// 调用牛顿法求解程序在分好的一小段上求解，将解存储在fn（j）
        f_coeff(j) = 2.0_DP / ( dble(n) * N1_Legendre(f_root(j)) * DN_Legendre(f_root(j)) ) !// 利用公式计算高斯点的权重
        write (*,'(1x,a,g0)') '高斯点序号: ', j
        write (*,'(1x,a,g0,2x,a,g0)') '高斯点: ', f_root(j), '高斯点权重', f_coeff(j)
        write(*,'(1x,a)') '------------------------------------------------------'
      End if
      m = m + h !// 执行完一次判断m向前推进一步
    End Do
  End subroutine root_coeff
  
  Real(Kind=DP) function func(x) !// 被积函数
    Implicit none
    Real(Kind=DP) :: x
    func = exp( -x*x/2.0_DP ) !// 每次计算在这里修改被积函数即可
  End function func

End module Gauss_Legendre

Program GaussianIntegral
  use Gauss_Legendre
  Implicit none
  Real (Kind=DP) :: f_root(n), f_coeff(n), x, a, b, answer
  Integer :: i
  Call root_coeff ( f_root, f_coeff ) !// 调用求高斯零点和权重的子函数
  a = -1.0_DP !// 积分上限
  b = 1.0_DP !// 积分下限   
  answer = 0.d0 !// 求积分结果赋初始值
  !// 一般区间[a,b]上的积分公式
  !// Integral[f(x),a,b] = Integral[f( ((b-a)*t+b+a)/2 )] * ( b-a )/2. t为N阶勒让德多项式的根
  Do i = 1, n
    answer = answer + f_coeff(i) * func( (a+b) / 2.0_DP + (b-a) / 2.0_DP * f_root(i) ) !// 高斯勒让德求积分公式     
  End Do
  answer = answer * (b-a) / 2.0_DP
  !// 精确解为1.71124878378430
  !// 数值解为1.711248783784299  
  Write(*,'(1x,a,g0)') '高斯-勒让德求积分结果: ', answer

End program GaussianIntegral

File: chap5/GaussianIntegral.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 22 16:29:18 2018

@author: luk
"""

import numpy as np
n = 11 # 设置求取高斯点的个数
eps = 1e-12  # 精度控制
a = -1.0; b = 1.0  # 积分区间

def N_Legendre(x):  # 生成n阶勒让德多项式
    global n
    
    a = np.zeros( (n,), dtype = np.float64 )
    a[0] = x  # 1阶勒让德多项式
    a[1] = 1.5 * x * x - 0.5  # 2阶勒让德多项式
    for i in range(2,n):
        # 利用递推关系产生n阶勒让德多项式
        a[i] = ( np.float64(i+i+1)*x*a[i-1] - np.float64(i)*a[i-2] ) / np.float64(i+1) 
    
    return a[n-1]

# print( N_Legendre(1.1) )
def N1_Legendre(x): # 生成n-1阶勒让德多项式
    global n
    
    a = np.zeros( (n,), dtype = np.float64 )
    a[0] = x
    a[1] = 1.5 * x * x - 0.5
    for i in range(2,n):
        a[i] = ( np.float64(i+i+1)*x*a[i-1] - np.float64(i)*a[i-2] ) / np.float64(i+1) 
        
    return a[n-2]
   
# print( N1_Legendre(1.1) )
    
def DN_Legendre(x): # 生成n阶勒让德多项式的导数表达式
    global n
    
    a = np.zeros( (n,), dtype = np.float64 )
    a[0] = x
    a[1] = 1.5 * x * x - 0.5
    for i in range(2,n):
        a[i] = ( np.float64(i+i+1)*x*a[i-1] - np.float64(i)*a[i-2] ) / np.float64(i+1) 
    
    result = ( a[n-2] - x*a[n-1] ) * n / ( 1.0 - x*x )
    
    return result
# print( DN_Legendre(1.1) )
    
def NewtonIteration( a, b ):  # 牛顿法求解函数的解
    nloop = 2000
    # a,b是传递进来的划分好的有一个解存在的区间
    x = ( a + b ) / 2.0
    i = 0
    while True:
        xtmp = x - N_Legendre(x) / DN_Legendre(x)  #  X(i+1) = Xi - f(Xi) / f'(Xi)  i = 1,2,...N
        i = i + 1
        if ( abs(xtmp-x) < eps and i > nloop ): break
        x = xtmp
    return x

f_root = np.zeros( (n,), dtype = np.float64 )  # 高斯点
f_coeff = np.zeros( (n,), dtype = np.float64 )  # 高斯系数

def root_coeff():  # 计算N阶勒让德多项式的根与去做权重系数
    global f_root, f_coeff
    global a, b
    
    h = 1e-6  # 步长
    j = 0  # 控制循环变量的初值
    m = -1.e0 - h 
    nstep = round( (b-a) / h )  # 此处的2为积分上限减去下限，积分区间为[-1,1]
    for i in range(nstep):
        if ( N_Legendre(m) * N_Legendre(m+h) < 0 ):
            f_root[j] = NewtonIteration( m, m+h )
            # 利用公式计算高斯点的权重
            f_coeff[j] = 2.0 / ( np.float64(n) * N1_Legendre(f_root[j]) * DN_Legendre(f_root[j]) )
            print( "高斯点序号:", j )
            print( "高斯点:", f_root[j], ",", "高斯点权重:", f_coeff[j] )
            print( "------------------------------------------------------- " )
            j = j + 1
        
        m = m + h  # 计算完后向前计算

# 计算高斯点与高斯权系数           
root_coeff()

# -----------------被积函数------------------
func = lambda x: np.exp( -x*x/2.0 )

# ------------------求积分-------------------
ans = 0.0
# 一般区间[a,b]上的求积分公式
# Integral[f(x),a,b] = Integral[f( ((b-a)*t+b+a)/2 )] * ( b-a )/2. t为N阶勒让德多项式的根
for i in range(n):
    ans = ans + f_coeff[i] * func( (a+b)/2.0 + (b-a)*f_root[i]/2.0 )
    
ans = ans * ( b-a ) / 2.0
print( "高斯-勒让德求积分结果:", ans )

print("please input NUMBER and stop!")
input()
